app/grandchallenge/publications/utils/__init__.py

```diff
@@ -1,11 +1,75 @@
+import re
+
 import requests
+from django.core.validators import RegexValidator
+from django.db import models
+from django.forms import CharField
 
-from grandchallenge.publications.models import (
-    PublicationType,
-    get_publication_type,
-)
 from grandchallenge.publications.utils.manubot import get_arxiv_csl
 
+# regex modified for python syntax from
+# https://www.crossref.org/blog/dois-and-matching-regular-expressions/
+DOI_REGEX = r"^10\.\d{4,9}/[-._;()/:a-z0-9]+$"
+ARXIV_REGEX = r"^\d{4}\.\d{4,5}$"
+
+identifier_validator = RegexValidator(regex=f"{DOI_REGEX}|{ARXIV_REGEX}")
+
+
+class PublicationType(models.TextChoices):
+    DOI = "D"
+    ARXIV = "A"
+
+
+class IdentifierField(models.CharField):
```

Comment:
Interesting stuff!

I think a better home for this would be in a `fields.py`. Would be great if you could move `get_publication_type`, `get_identifier_csl` and `get_publication_url` to `@properties` on the field now, so we could do things like `publication.identifier.kind` (note not type as that is reserved), `publication.identifier.csl` and `publication.identifier.url` rather than using separate methods.

Reply:
Hm, I cannot get it to work like this. If I add a property to the field and then try to call `publication.identifier.kind`, for example, it throws an error saying that a string doesn't have an attribute 'kind'. The property is registered on the Field class, but I can't call it.

Reply:
Ah yes, my mistake, you will need to return and take a different class with those properties, rather than a string. Worth reading through this documentation https://docs.djangoproject.com/en/4.0/howto/custom-model-fields/ and especially https://docs.djangoproject.com/en/4.0/howto/custom-model-fields/#converting-values-to-python-objects-1

Reply:
Sorry, I'm still stuck. I created a new class `PublicationIdentifier` with those properties and I overwrite the `from_db_value` and `to_python` methods on the custom field to return an instance of this new class. That seems to work, but now saving a new `Publication` to the database does not work. It can't adapt the type 'PublicationIdentifier', but I haven't been able to figure out what I'm missing.

Reply:
It took forever, but I think I got it now.